mpitests/test_expval.py

```diff
@@ -22,114 +22,260 @@
+        dev_mpi = qml.device(
+            "lightning.gpu", wires=num_wires, mpi=True, c_dtype=C_DTYPE, batch_obs=batch_obs
+        )
+
+        def circuit():
+            qml.RX(0.4, wires=[0])
+            qml.RY(-0.2, wires=[numQubits - 1])
+            return qml.expval(ham)
 
-        O1 = qml.Hadamard(wires=[0])
-        O2 = qml.Hadamard(wires=[1])
+        cpu_qnode = qml.QNode(circuit, dev_cpu)
+        mpi_qnode = qml.QNode(circuit, dev_mpi)
 
-        dev.apply(
-            [qml.RY(theta, wires=[0]), qml.RY(phi, wires=[1]), qml.CNOT(wires=[0, 1])],
-            rotations=[*O1.diagonalizing_gates(), *O2.diagonalizing_gates()],
+        assert np.allclose(cpu_qnode(), mpi_qnode(), atol=tol, rtol=0)
+
+    def test_expval_non_pauli_word_hamiltionian(self, tol, C_DTYPE, batch_obs):
+        """Tests expectation values of non-Pauli word Hamiltonians."""
+        dev_mpi = qml.device(
+            "lightning.gpu", wires=3, mpi=True, c_dtype=C_DTYPE, batch_obs=batch_obs
         )
+        dev_cpu = qml.device("lightning.qubit", wires=3)
+
+        theta = 0.432
+        phi = 0.123
+        varphi = -0.543
+
+        def circuit():
+            qml.RX(theta, wires=[0])
+            qml.RX(phi, wires=[1])
+            qml.RX(varphi, wires=[2])
+            qml.CNOT(wires=[0, 1])
+            qml.CNOT(wires=[1, 2])
+            return qml.expval(0.5 * qml.Hadamard(2))
+
+        cpu_qnode = qml.QNode(circuit, dev_cpu)
+        mpi_qnode = qml.QNode(circuit, dev_mpi)
+
+        assert np.allclose(cpu_qnode(), mpi_qnode(), atol=tol, rtol=0)
 
-        res = np.array([dev.expval(O1), dev.expval(O2)])
-        expected = np.array(
-            [
-                np.sin(theta) * np.sin(phi) + np.cos(theta),
-                np.cos(theta) * np.cos(phi) + np.sin(phi),
-            ]
-        ) / np.sqrt(2)
-        assert np.allclose(res, expected, tol)
-
-    @pytest.mark.parametrize("n_wires", range(1, 8))
-    def test_hermitian_expectation(self, n_wires, theta, phi, tol):
+    @pytest.mark.parametrize("theta, phi", list(zip(THETA, PHI)))
+    @pytest.mark.parametrize("n_wires", range(1, numQubits))
+    def test_hermitian_expectation(self, n_wires, theta, phi, tol, C_DTYPE, batch_obs):
         """Test that Hadamard expectation value is correct"""
-        n_qubits = 7
+        n_qubits = numQubits - 1
         dev_def = qml.device("default.qubit", wires=n_qubits)
-        dev = qml.device(device_name, mpi=True, wires=n_qubits)
+        dev = qml.device(
+            device_name, mpi=True, wires=n_qubits, c_dtype=C_DTYPE, batch_obs=batch_obs
+        )
         comm = MPI.COMM_WORLD
 
         m = 2**n_wires
         U = np.random.rand(m, m) + 1j * np.random.rand(m, m)
         U = U + np.conj(U.T)
-        U = U.astype(dev.C_DTYPE)
+        U = U.astype(dev.c_dtype)
         comm.Bcast(U, root=0)
         obs = qml.Hermitian(U, wires=range(n_wires))
 
         init_state = np.random.rand(2**n_qubits) + 1j * np.random.rand(2**n_qubits)
```

Comment:
Do we want to replace it with `np.lingalg`?

Reply:
Done!